serverless/serverless
# Only re-deploy when code or service updated

## What did you implement:

Closes #2710

Checks if the function code or service configuration has changed and only deploys if that's the case.

## How did you implement it:

Add a `sha256` based hash comparison for the files in question (e.g. the `.zip` or `CloudFormation` files). Only deploy when the hashes are different.

This PR introduces a new AWS related lifecycle event (`checkForChanges`) and sets a globally available flag (`this.serverless.service.provider.shouldNotDeploy`) so that others can use the result of the check as well. Thanks for pointing that out @HyperBrain 👍 

## How can we verify it:

Use the following service:

```yml
service:
  name: my-service

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
  goodbye:
    handler: handler.goodbye
```

1. Run `serverless deploy` to deploy everything.
2. Run `serverless deploy function -f hello` --> Console will tell you that nothing was deployed because the code was not changed.
3. Update `hello` function code
4. Run `serverless deploy function -f hello` --> New function code will be uploaded
5. Run `serverless deploy`
6. Run `serverless deploy` (yes, a second time) --> Nothing will be updated since nothing changed
7. Change the `hello` function code
8. Run `serverless deploy` --> Deployment should be triggerend since the function code changed
9. Run `serverless deploy` --> Nothing should change since nothing was updated
10. Change configuration by e.g. adding a new function
11. Run `serverless deploy` --> Deployment should succeed

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
